_posts/2019-07-26-pysam-with-s3.md

```diff
@@ -0,0 +1,182 @@
+1) https://www.biostars.org/p/213305/
+2) https://github.com/pysam-developers/pysam/issues/557
+3) https://www.biostars.org/p/213305/ 
+
+Although several sources claim that accessing s3 objects should just work, I have not been able to find a single worked example. After many days of digging on the web and coming across several problems, my team and I figured out how to successfully use PySAM and htslib to access private S3 bucket; we describe how below.
+
+### 1) Download and install htslib version from developer branch
+   
+We believe the root of the problem is AWS's move to a [new algorithm](https://docs.aws.amazon.com/AmazonS3/latest/API/sig-v4-authenticating-requests.html) for signed URL access to private objects which is not supported by the current release of htslib (v1.9); but is supported on the `devel` branch.  To use this new version, the installation instructions can be found here: https://github.com/samtools/htslib . Make sure to add the --enable-s3 option during installation. 
+
+To check whether htslib installed correctly, use the command:
+```
+which htsfile
+```
+The output should be:
+```
+/usr/local/bin/htsfile
+```
+
+### 2) Open file from AWS S3 path with htslib and environment variables
+We can demonstrate that a file from a S3 path can be accessed with the new htslib. 
+
+##### Set environment variables
+```
+AWS_SECRET_ACCESS_KEY=XXXXXXXX
+AWS_ACCESS_KEY_ID=XXXXXXXX
+```
+##### Run the command
+```
+htsfile s3://1000genomes/1000G_2504_high_coverage/data/ERR3243163/NA12546.final.cram
+```
+
+### 3) Open file from non AWS implementations of S3 like minIO with htslib
+Accessing non AWS implementations of S3 like minio has a different URL format, and require a ./s3cfg file instead of setting environment variables. We will use the testing server play.min.io:9000 and we created our own test bucket called testfiles with the file NA18537.vcf.gz
+
+##### Create ./s3cfg file instead of using environment variables
+We created ./s3cfg file in the root directory of your computer. The file should look like this:
+```
+[default]
+access_key = Q3AM3UQ867SPQQA43P2F  
+secret_key = zuf+tfteSlswRu7BJ86wekitnifILbZam1KYY3TG  
+host_base = min.io:9000
+```
+
+If you're wondering why the host_base is min.io:9000 and not play.min.io:9000, it is because htslib reads the host_base, and the file path in a specific format, and we had to play around with it for awhile to get it to read the host name in the way we wanted. When using AWS, the host base is known to be e.g. "s3.aws.amazon.com". In general, when calling the s3 path with htslib, the path is in the following format:
```

Comment:
This explanation could use some work.

So first, we can't set host base via environment variables, so we have to use a fallback method for configuration, the .s3cfg file, as above.   

But we also have to lie a bit about the bucket name.  The URL for accessing an object named  say "ERR3243163/NA12546.final.cram" in a bucket called 1000genomes on s3 via https looks like this:

https://1000genomes.s3.amazonaws.com/ERR3243163/NA12546.final.cram

e.g. bucketname.hostbase/object

and that's what the s3 code in htslib assumes is the structure of the URL.

But using (say) play.min.io the URL structure is::

play.min.io:9000/prefix/object

So we need to pretend "play" is the bucket here and the host_base is min.io:9000.

Note that s3cfg allows you to define different profiles, so you could access multiple servers; here we're just using one profile, called 'default', and using that as s3://[profile]@[bucket]//... ; we could just as easily have called it something else